lib/Sema/CSDiag.cpp

```diff
@@ -6508,7 +6554,75 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
         }
         return true;
       }
-      
+
+      if (inferredArgCount == 1 && actualArgCount > 1) {
+        // Let's see if inferred argument is actually a tuple inside of Paren.
+        if (auto *argTupleTy = inferredArgType->getAs<TupleType>()) {
+          // Looks like the number of closure parameters matches number
+          // of inferred arguments, which means we can warn about SE-0110
+          // and provide a proper fix-it.
+          if (argTupleTy->getNumElements() == actualArgCount) {
+            auto diag = diagnose(params->getStartLoc(),
+                                 diag::closure_parameter_tuple_destructuring,
+                                 argTupleTy);
+
+            Type actualArgType;
+            if (auto *actualFnType = CE->getType()->getAs<AnyFunctionType>())
+              actualArgType = actualFnType->getInput();
+
+            auto typeComparator = [&](Type lhs, Type rhs) -> bool {
+              auto *typeVar = lhs->getAs<TypeVariableType>();
+              if (!typeVar)
+                return lhs->isEqual(rhs);
+
+              // Type variables are allowed only in the actual arguments.
+              if (rhs->getKind() == TypeKind::TypeVariable)
+                return false;
+
+              // If there is an inferred type associated with closure
+              // parameter, let's check that against inferred parameter type.
+              if (auto fixedType = CS->getFixedType(typeVar))
+                return fixedType && fixedType->isEqual(rhs);
+
+              // type variabls in this case represent types of
+              // the closure parameters and we only check structure
+              // of the type itself.
+              return true;
+            };
+
+            // Can't produce fix-it, because inferred parameter
+            // and actual argument type structure are not the same.
+            if (!matchArgStructure(actualArgType, argTupleTy, typeComparator))
+              return true;
+
+            auto &manager = CS->getASTContext().SourceMgr;
+            auto paramRange = CharSourceRange(
+                manager, params->getStartLoc(),
+                Lexer::getLocForEndOfToken(manager, params->getEndLoc()));
+
+            SmallString<64> fixIt;
+            llvm::raw_svector_ostream OS(fixIt);
+
+            // If this is multi-line closure we'd have to insert new lines
+            // in the suggested 'let' to keep the structure of the code in
+            // tact, otherwise just use ';' to keep everything on the same
+            // line.
+            if (!CE->hasSingleExpressionBody())
```

Comment:
`hasSingleExpressionBody` is not the same condition is "is written on multiple lines". I think it's worth really looking for a newline to decide whether to insert newlines.

On the other hand, `hasSingleExpressionBody` *does* indicate that the body was previously available for type inference and now it won't be, and so we should include the parameter type in the fix-it.